pingrid.py

```diff
@@ -448,13 +451,20 @@ def parse_colormap(s: str) -> np.ndarray:
     #     len(vs),
     #     [f"{v.red:02x}{v.green:02x}{v.blue:02x}{v.alpha:02x}" for v in vs],
     # )
-    rs = np.array([vs[int(i / 256.0 * len(vs))] for i in range(0, 256)], np.uint8)
+    if thresholds is None:
+        rs = np.array([vs[int(i / 256.0 * len(vs))] for i in range(0, 256)], np.uint8)
+    else:
+        rs = np.full([256, 4], np.nan)
+        for i in range(0, thresholds.size):
+            rs[int(255.*(thresholds[i]-thresholds[0])/(thresholds[-1]-thresholds[0])), :] = np.array([vs[i]], np.uint8)
+        dfrs = pd.DataFrame(rs)
+        rs = dfrs.interpolate().astype(int).values
```

Comment:
Is this a python version of existing ingrid colorscale structure? Ingrid puts the color values in the cf metadata, and xarray loads it into ds.attrs. Do you know if it does the same for the thresholds? If it does, then we should parse the ingrid representation as we do for the color values.

Reply:
Do you have a link handy to an ingrid dataset whose colorscale is defined using thresholds like this?

Reply:
I think/hope I understand now.

`parse_colormap` was originally written to parse the `colormap` attribute of datasets downloaded from Ingrid. In Ingrid's representation, the colormap you defined in your last link (adapted from the tutorial) looks like this:

[ 9470064 12632256 8388608 [ 15570276 26 ] [ 16777184 89 ] [ 13826810 26 ] [ 42495 38 ] [ 7504122 51 ] 6053069 [ 6053069 25 ] 12632256 ]

([source](https://iridl.ldeo.columbia.edu/SOURCES/.NOAA/.NCEP/.CPC/.CAMS/.anomaly/.temp/startcolormap/-5./5./RANGE/SlateGrey/grey/NavyBlue/-5./VALUE/CornflowerBlue/-4./VALUE/LightCyan/-0.5/VALUE/LightGoldenrodYellow/0.5/VALUE/orange/2/VALUE/salmon/4./VALUE/IndianRed/IndianRed/5./bandmax/grey/endcolormap/colormap/#expert))

That representation plus the min and max are sufficient to map values to colors. We don't need to keep track of the thresholds explicitly; they are implicit in the way the quanta 1-255 have been mapped to colors.

The problem in this case is that you want to set the thresholds on the fly, not ahead of time when the dataset is downloaded from Ingrid. It's not practical to have Ingrid generate the colormap on the fly, so you want to duplicate Ingrid's logic in python. This seems reasonable to me, but I don't think `parse_colormap` is the place to do it. `parse_colormap` was intended to parse Ingrid's output. If we're going to have a function that replaces this Ingrid functionality, it should be a separate function. Otherwise, there's an interaction between thresholds that were defined in Ingrid and are implicit in the `s` value and thresholds that are explicit in the `thresholds` value that is too complicated for me to get my head around.

Am I making any sense yet?

Reply:
yes totally and that clears things up for me too. So the confusion originated from:

1. my assumption to start building more colormaps stuff on what was in pingrid, which I didn't get a sense that it had a specific Ingrid parsing purpose;
2. Not having fully understood the representation above;
3. having to deal with the one and only sophisticated case of a dynamic colorscale.

And thus it looks like what I did is create some objects in Python that have the looks of what we get from Ingrid so that I can use the tools that make a Python colorscale from Ingrid inputs... which is dumb. 

So I guess I need to pay more attention to what exactly creates a colorscale and from what input; then use that and leave parse_colormap alone (and return it to its original form). Unless you have some more inputs you'd like to bring in but my understanding is that your reflection on the multiple tools that allows to make the colorscales is just starting.

Reply:
Yes, it seems we need a python version of Ingrid's startcolormap ... endcolormap.

Consider using a static colormap for now and leaving the dynamic version for a subsequent PR, which I could help with.

Reply:
@aaron-kaplan , how do you get that structure?
[ 9470064 12632256 8388608 [ 15570276 26 ] [ 16777184 89 ] [ 13826810 26 ] [ 42495 38 ] [ 7504122 51 ] 6053069 [ 6053069 25 ] 12632256 ]
To create my colorscales, I found a website that translated RGB to those numbers. And for the thresholds I used the thresholds attribute that I understand now is not needed. Did you calculate the 1-255 numbers manually?